### PR TITLE
SCMOD-8752: Split autoscale back off

### PR DESCRIPTION
**Pull request by Anthony-McGreevy**
**Jira**: https://portal.digitalsafe.net/browse/SCMOD-8752
**Build**: http://sou-jenkins2.swinfra.net/job/Autoscaler/view/Developer/job/Autoscaler~autoscaler~SCMOD-8752~CI/

### CODE REVIEW

File: docs/pages/en-us/Getting-Started.md [hunk 1]
```diff
@@ -68,6 +68,8 @@ Services deployed by Marathon (such as Example-Worker) contain **labels** in the
 `autoscale.maxinstances`: The maximum number of instances a worker can have; set as your level of resources and quality of service dictate.<br>
 `autoscale.interval`: Measured in seconds. This controls how fervently the autoscaler monitors the service (i.e. how often it re-evaluates the number of services required to service the workload).<br>
 `autoscale.backoff`: Manages the number of intervals to skip after a scale up or down command is issued while monitoring. This prevents unusual values being considered when the system is in an unstable state.
+`autoscale.scaledownbackoff`: Optional. Manages the number of intervals to skip after a scale down command is issued while monitoring. This prevents unusual values being considered when the system is in an unstable state. If this is not set it will fall back to `autoscale.backoff`.
+`autoscale.scaleupbackoff`: Optional. Manages the number of intervals to skip after a scale up command is issued while monitoring. This prevents unusual values being considered when the system is in an unstable state. If this is not set it will fall back to `autoscale.backoff`.
```
Comment (dermot-hardy) [inline docs/pages/en-us/Getting-Started.md]: You haven't put the `<br>`s like the other ones.

File: autoscale-core/src/main/java/com/hpe/caf/autoscale/core/ScalerThread.java [hunk 2]
```diff
@@ -266,4 +290,33 @@ private int establishMemLimitReached(final double currentMemoryLoad)
...
+        }
+        return false;
+    }
+
+    private boolean defaultBackoff()
+    {
+        if (backoffCount > backoffAmount) {
+            backoff = false;
+            backoffCount = 0;
+            return false;
```
Comment (dermot-hardy) [inline autoscale-core/src/main/java/com/hpe/caf/autoscale/core/ScalerThread.java:322]: The logic doesn't seem right to me.  Why does these variable resets need to be done in the default case but not in the new cases you've added?

**Review by dermot-hardy — CHANGES_REQUESTED** [review]

File: autoscale-core/src/main/java/com/hpe/caf/autoscale/core/ScalerThread.java [hunk 3]
```diff
@@ -266,4 +290,36 @@ private int establishMemLimitReached(final double currentMemoryLoad)
...
+                }
+                default: {
+                    shouldBackoff = defaultBackoff();
+                    break;
+                }
+            }
+        }
+        if (shouldBackoff) {
+                    backoff = false;
+        backoffCount = 0;
```
Comment (dermot-hardy) [inline autoscale-core/src/main/java/com/hpe/caf/autoscale/core/ScalerThread.java]: Indenting screwed up here.

File: autoscale-core/src/main/java/com/hpe/caf/autoscale/core/ScalerThread.java [hunk 4]
```diff
@@ -266,4 +290,36 @@ private int establishMemLimitReached(final double currentMemoryLoad)
...
+                    break;
+                }
+                default: {
+                    shouldBackoff = defaultBackoff();
+                    break;
+                }
+            }
+        }
+        if (shouldBackoff) {
+                    backoff = false;
```
Comment (dermot-hardy) [inline autoscale-core/src/main/java/com/hpe/caf/autoscale/core/ScalerThread.java]: I don't get it.  Why would you be setting `backoff` to `false` when `shouldBackoff` is `true`?  Won't this stop the backing off?

**Review by dermot-hardy — CHANGES_REQUESTED** [review]

**Review by dermot-hardy — CHANGES_REQUESTED** [review]
I'm sorry but the logic still doesn't look right to me.

**Comment by Anthony-McGreevy** [comment]
Tested changes on larry and seen the modified scale up and scale down backoffs being set on the scaler threads for different apps, test carried out with 21,000 docs.

File: autoscale-core/src/main/java/com/hpe/caf/autoscale/core/ScalerThread.java [hunk 5]
```diff
@@ -117,10 +139,10 @@ private void handleAnalysis()
             LOG.debug("Performing scaling checks for service {}", serviceRef);
             action = analyser.analyseWorkload(instances);
             LOG.debug("Workload Analyser determined that the autoscaler should {} {} by {} instances",
-                     action.getOperation(), serviceRef, action.getAmount());
+                      action.getOperation(), serviceRef, action.getAmount());
```
Comment (dermot-hardy) [inline autoscale-core/src/main/java/com/hpe/caf/autoscale/core/ScalerThread.java]: Pointless whitespace change

File: autoscale-core/src/main/java/com/hpe/caf/autoscale/core/ScalerThread.java [hunk 6]
```diff
@@ -117,10 +139,10 @@ private void handleAnalysis()
...
             action = governor.govern(serviceRef, action, currentMemoryLimitStage);
             LOG.debug("Governor determined that the autoscaler should {} {} by {} instances",
-                     action.getOperation(), serviceRef, action.getAmount());
+                      action.getOperation(), serviceRef, action.getAmount());
```
Comment (dermot-hardy) [inline autoscale-core/src/main/java/com/hpe/caf/autoscale/core/ScalerThread.java]: Pointless whitespace change

File: autoscale-core/src/main/java/com/hpe/caf/autoscale/core/ScalerThread.java [hunk 7]
```diff
@@ -254,7 +278,7 @@ private void handleAlerterDispatch(final double memLoad) throws ScalerException
             alertDispatcher.dispatchAlert(emailBody);
         }
     }
-    
+
```
Comment (dermot-hardy) [inline autoscale-core/src/main/java/com/hpe/caf/autoscale/core/ScalerThread.java]: Pointless whitespace change

File: autoscale-core/src/main/java/com/hpe/caf/autoscale/core/ScalerThread.java [hunk 8]
```diff
@@ -266,4 +290,37 @@ private int establishMemLimitReached(final double currentMemoryLoad)
         }
         return 0;
     }
+
+    private boolean backoffComplete()
```
Comment (dermot-hardy) [inline autoscale-core/src/main/java/com/hpe/caf/autoscale/core/ScalerThread.java]: The function naming is a little confusing here.  It seems strange to have an `isBackoffComplete()` a) without the `is` and b) even though there may never have been any backoff.

**Review by dermot-hardy — APPROVED** [review]

**Pull request merged**